File: Interpreter/Lexer.mjs
function tokenize(code) {
    let tokens = [];

    function addToken(type, value) {
        tokens.push({type, value})
    }

    for (let i = 0; i < code.length; i++) {
        let char = code[i];

        if (/\s/.test(char)) continue;

        // Parens
        if (char === '(' || char === ')') {
            addToken('paren', char);
        }

        // String
        else if (char === '"') {
            let value = '';
            char = code[++i];

            while (char !== '"') {
                value += char;
                char = code[++i];
            }

            char = code[++i];
            i--;

            addToken('str', value);
        }

        // Number
        else if (/\d/.test(char)) {
            let value = char;
            char = code[++i];

            while (/\d/.test(char) || char === '.') {
                value += char;
                char = code[++i];
            }
            i--;

            addToken('num', value);
        }

        // Operators
        else if (char === '+' || char === '-' ||
            char === '*' || char === '/' ||
            char === '%' || char === '^' ||
            char === '!' || char === '-') {
            addToken('op', char);
        }
    }

    return tokens
}

export default {tokenize};

File: src/lexer.rs
#[derive(Debug, PartialEq)]
pub enum Token {
    LParen,
    RParen,
    LBrace,
    RBrace,
    Number(i32),
    String(String),
    Symbol(String),
}

pub fn lex(input: String) -> Vec<Token> {
    let mut tokens = Vec::new();

    let sanitized_input = &input
        .replace("(", " ( ")
        .replace(")", " ) ")
        .replace("{", " { ")
        .replace("}", " } ");

    for word in sanitized_input.split_ascii_whitespace() {
        let token = match word {
            "(" => Some(Token::LParen),
            ")" => Some(Token::RParen),
            "{" => Some(Token::LBrace),
            "}" => Some(Token::RBrace),
            _ => {
                if let Ok(num) = word.parse::<i32>() {
                    Some(Token::Number(num))
                } else if word.starts_with('"') {
                    Some(Token::String(word[1..word.len() - 1].to_string()))
                } else {
                    Some(Token::Symbol(word.to_string()))
                }
            }
        };
        tokens.push(token.unwrap());
    }

    tokens
}

#[cfg(test)]
mod tests {
    use crate::lexer::lex;
    use crate::lexer::Token::*;

    #[test]
    fn test_lexer() {
        let result = lex("(+ 2 24)".to_string());
        assert_eq!(
            result,
            vec![
                LParen,
                Symbol("+".to_string()),
                Number(2),
                Number(24),
                RParen,
            ]
        );
    }

    #[test]
    fn test_lexer2() {
        let result = lex(r#"
        (fun {areacircle rad} { do
            (print "Radius:\t" rad)
            (print "Area:\t" (* rad rad))
        })
        "#
        .to_string());
        assert_eq!(
            result,
            vec![
                LParen,
                Symbol("fun".to_string()),
                LBrace,
                Symbol("areacircle".to_string()),
                Symbol("rad".to_string()),
                RBrace,
                LBrace,
                Symbol("do".to_string()),
                LParen,
                Symbol("print".to_string()),
                String("Radius:\\t".to_string()),
                Symbol("rad".to_string()),
                RParen,
                LParen,
                Symbol("print".to_string()),
                String("Area:\\t".to_string()),
                LParen,
                Symbol("*".to_string()),
                Symbol("rad".to_string()),
                Symbol("rad".to_string()),
                RParen,
                RParen,
                RBrace,
                RParen,
            ]
        );
    }
}
